packages/styled-base/types/tests.tsx

```diff
@@ -296,3 +310,73 @@ const StyledFCWithDefaultProps = styled(FCWithDefaultProps)`
   background-color: red;
 `
 const fcInstance = <StyledFCWithDefaultProps />
+
+interface PropsA {
+  title: string
+}
+interface PropsB {
+  content: string
+}
+
+type PropsAB = PropsA | PropsB
+class A extends React.Component<PropsAB> {
+  static getDerivedStateFromProps(props: PropsAB) {
+    return null
+  }
+
+  render() {
+    return null
+  }
+}
+
+const B = styled(A)`
+  color: red;
+`
+;<B title="test" />
+;<B content="test" />
+// Because these are not tagged unions, they are not mutually exclusive, we can do both
+;<B title="test" content="test" />
+
+interface TaggedPropsA {
+  tag: 'a'
+  title: string
+}
+interface TaggedPropsB {
+  tag: 'b'
+  content: string
+}
+type TaggedPropsAB = TaggedPropsA | TaggedPropsB
+class C extends React.Component<TaggedPropsAB> {
+  render() {
+    return null
+  }
+}
+
+const D = styled(C)`
+  color: red;
+`
+;<D tag="a" title="test" />
```

Comment:
Would it be possible having such `tag` being any component and having other acceptable props being based on it?  This would probably require conditional types - but maybe we could cover `as` prop?

Reply:
Not sure that is possible. It would require the type to change based on one of it's keys.

Reply:
Let's not try it out for now then 😉